### PR TITLE
Implement support for kerberos authentication

### PR DESCRIPTION
**Pull request by tsipinakis**
## Please describe the change you are making

Implement a kerberos authentication flow. It includes an additional webhook to the authentication server for authorization. 
The authentication method requires the keytab to be mounted in the container (or be present on the host). 

Example configuration:
```
    auth:
      method: kerberos
      kerberos:
        keytab: '/etc/krb5.keytab'
        authz:
          enable: true
          url: http://127.0.0.1:8080
```

In order to run this, the changes from https://github.com/ContainerSSH/agent/pull/3 and https://github.com/ContainerSSH/gokrb5/pull/2

TODO:
- [ ] Test/Add support for docker backend
- [x] Add `principalUser` to authz request in case ticket username != ssh username
- [ ] Finish off rough edges
    - [x] ConnectionId/Remote address not being sent to authz server
    - [ ] ...
- [ ] Documentation, look into all the files that need updating
- [ ] Tests

## Are you the owner of the code you are sending in, or do you have permission of the owner?

Sent with permission of the owner

## The code will be published under the MIT-0 license. Have you read and understood this license?

Yes

### CODE REVIEW

File: internal/auth/client_kerberos.go [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+		return err
+	}
+
+	// 50 = SSH_MSG_USERAUTH_REQUEST
+	if field.Request != 50 || field.Service != "ssh-connection" || field.Method != "gssapi-with-mic" {
+		return fmt.Errorf("Invalid MIC field")
+	}
+
+	if k.client.enforceUsername && field.UserName != k.username {
+		return fmt.Errorf("Invalid username for ticket")
```
Comment (janosdebugs) [inline internal/auth/client_kerberos.go:311]: Please use the `message.UserMessage` to produce a user-friendly error message.

File: internal/auth/client_kerberos.go [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+
+	authzRequest := auth.AuthorizationRequest{
+		PrincipalUsername: k.username,
+		LoginUsername: username,
+		RemoteAddress: k.remoteAddr.String(),
+		ConnectionID: k.connectionId,
+	}
+	authzResponse := &auth.ResponseBody{}
+
+	statusCode, err := k.client.http.Post("/authz", authzRequest, authzResponse)
```
Comment (janosdebugs) [inline internal/auth/client_kerberos.go:336]: This code should be in a retry and should be properly logged. See `processAuthWithRetry` for an example.

File: internal/auth/client_kerberos.go [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+	if err != nil {
+		k.err = err
```
Comment (janosdebugs) [inline internal/auth/client_kerberos.go:338]: Please consider wrapping this error.

File: internal/auth/client_kerberos.go [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+		k.success = false
+		return err
+	}
+	if statusCode != 200 {
+		k.err = fmt.Errorf("Non 200 error code from authz server")
```
Comment (janosdebugs) [inline internal/auth/client_kerberos.go:343]: Please use `message.WrapUser`

File: config/auth.go [hunk 5]
```diff
@@ -293,3 +299,44 @@ func (o *AuthOIDCConfig) Validate() error {
...
+	// will be written inside the user containers. For kerberos to work
+	// correctly this should match the contents of the KRB5CC environment
+	// variable.
+	CredentialCachePath string `json:"credentialCachePath" yaml:"credentialCachePath" default:"/tmp/krb5cc"`
+	// AllowPassword controls whether kerberos-based password
+	// authentication should be allowed. If set to false only GSSAPI
+	// authentication will be permitted
+	AllowPassword bool `json:"allowPassword" yaml:"allowPassword" default:"true"`
+	// ConfigPath is the path of the kerberos configuration file. This is only used for password authentication.
+	ConfigPath string `json:"configPath" yaml:"configPath" default:"/etc/krb5.conf"`
```
Comment (janosdebugs) [inline config/auth.go:333]: I would consider setting this to `/etc/containerssh/krb5.conf` so the container version automatically picks it up from the config directory.

File: internal/auth/client_kerberos_factory.go [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+	kt, err := keytab.Load(cfg.Kerberos.Keytab)
+	if err != nil {
+		return nil, err
+	}
+
+	var conf *krb5cfg.Config
+	if cfg.Kerberos.AllowPassword {
+		conf, err = krb5cfg.Load(cfg.Kerberos.ConfigPath)
+		if err != nil {
+			return nil, err
```
Comment (janosdebugs) [inline internal/auth/client_kerberos_factory.go:38]: Please wrap this error for user friendliness.

File: internal/auth/client_kerberos_factory.go [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+		return nil, fmt.Errorf("authentication is not set to kerberos")
+	}
+
+	if err := cfg.Validate(); err != nil {
+		return nil, err
+	}
+
+	kt, err := keytab.Load(cfg.Kerberos.Keytab)
+	if err != nil {
+		return nil, err
```
Comment (janosdebugs) [inline internal/auth/client_kerberos_factory.go:31]: Please wrap this error for user friendliness and add the path to the keytab file to the error message.

File: internal/auth/handler.go [hunk 8]
```diff
@@ -34,4 +34,11 @@ type Handler interface {
 		RemoteAddress string,
 		ConnectionID string,
 	) (bool, map[string]string, error)
+
+	OnAuthorization(
```
Comment (janosdebugs) [inline internal/auth/handler.go:38]: Please add a comment to this function and explain how to use this. I would also consider renaming it to make it explicit that this is for Kerberos.

File: internal/kubernetes/networkHandler.go [hunk 9]
```diff
@@ -87,6 +89,44 @@ func (n *networkHandler) OnHandshakeSuccess(username string, clientVersion strin
 		if n.pod, err = n.cli.createPod(ctx, n.labels, n.annotations, env, nil, nil); err != nil {
 			return nil, err
 		}
+		if b64creds, ok := metadata["KRB5CCDATA"]; ok {
+			_, ok := metadata["KRB5CCPATH"]
+			if !ok {
+				return nil, fmt.Errorf("Provided credentials with no path")
```
Comment (janosdebugs) [inline internal/kubernetes/networkHandler.go:95]: Please use `message.UserMessage` to produce a user friendly error message.

File: internal/kubernetes/networkHandler.go [hunk 10]
```diff
@@ -87,6 +89,44 @@ func (n *networkHandler) OnHandshakeSuccess(username string, clientVersion strin
 		if n.pod, err = n.cli.createPod(ctx, n.labels, n.annotations, env, nil, nil); err != nil {
 			return nil, err
 		}
+		if b64creds, ok := metadata["KRB5CCDATA"]; ok {
```
Comment (janosdebugs) [inline internal/kubernetes/networkHandler.go:92]: The Kubernetes backend shouldn't know anything about Kerberos. This way we'll have to implement Kerberos handling in all backends separately. Furthermore, the metadata can be exposed to the container as env variables anyway. I would suggest adding this logic to the agent as that will result in a much cleaner implementation. That would make it possible to pass this through via the SSH proxy too, not just for container backends.

If that doesn't work we need to make an API for deploying files, which I'm not a fan of because it will introduce added complexity. Furthermore, it will be difficult to make sure that in the future this doesn't escalate into all kinds of shenanigans.

File: internal/kubernetes/networkHandler.go [hunk 11]
```diff
@@ -87,6 +89,44 @@ func (n *networkHandler) OnHandshakeSuccess(username string, clientVersion strin
...
+				return nil, fmt.Errorf("Provided credentials with no path")
+			}
+
+			credCtx, cancelFunc := context.WithTimeout(
+				context.Background(),
+				n.config.Timeouts.CommandStart,
+			)
+			defer cancelFunc()
+
+			exec, err := n.pod.createExec(credCtx, []string{n.config.Pod.AgentPath, "write-file", "/tmp/krb5cc"}, map[string]string{}, false)
```
Comment (janosdebugs) [inline internal/kubernetes/networkHandler.go:104]: This does not seem to handle the `session` case, only the `connection` mode. This will cause problems for anyone running in `session` mode.

File: internal/auth/handler.go (hunk 8 above)
Comment (tsipinakis) [reply internal/auth/handler.go:38]: This is deliberate. I was thinking other auth backends (i.e. GitHub ?) can use this for authz - this also solves the `EnforceUsername` problems of the GitHub backend if this is added to that as well.

Comment (janosdebugs) [reply internal/auth/handler.go:38]: Works for me. Let's document accordingly.

File: config/auth.go [hunk 12]
```diff
@@ -293,3 +299,44 @@ func (o *AuthOIDCConfig) Validate() error {
...
+	}
+	return nil
+}
+
+type AuthKerberosClientConfig struct {
+	// Keytab is the path to the kerberos keytab
+	Keytab string `json:"keytab" yaml:"keytab" default:"/etc/krb5.keytab"`
+	// EnforceUsername specifies whether to check that the username of the
+	// authenticated user matches the SSH username entered. ...
+	EnforceUsername bool `json:"enforceUsername" yaml:"enforceUsername" default:"true"`
```
[the remark this answers is not in the document]
Comment (tsipinakis) [reply config/auth.go:330]: Confirmed that disabling this does indeed work.

File: internal/auth/client_kerberos.go [hunk 13]
```diff
@@ -0,0 +1,366 @@
...
+
+		a := st.APReq
+		skew, err := time.ParseDuration("5m")
+		if err != nil {
+			panic(err)
+		}
+
+		hostAddr := types.HostAddressFromNetIP(k.remoteAddr)
+
+		// TODO: Add option for strict acceptor check
```
[the remark this answers is not in the document]
Comment (tsipinakis) [reply internal/auth/client_kerberos.go]: Added :)

File: internal/auth/client_kerberos.go [hunk 14]
```diff
@@ -0,0 +1,366 @@
...
+	st.Settings = krb5svc.NewSettings(k.client.keytab)
+	verified, _ := st.Verify()
+
+	if verified {
+		ctx := st.Context()
+		id := ctx.Value(spnego.CtxCredentials).(goidentity.Identity)
+		k.username = id.UserName()
+
+		a := st.APReq
+		skew, err := time.ParseDuration("5m")
```
[the remark this answers is not in the document]
Comment (tsipinakis) [reply internal/auth/client_kerberos.go]: Added as a configuration option, though not sure in what context one would need to change it.

File: internal/auth/client_kerberos.go [hunk 15]
```diff
@@ -0,0 +1,477 @@
...
+	// UserName is the username that the user requests to log in as
+	UserName string
+	// Service is the service being used ('ssh-connection')
+	Service string
+	// Method is the authentication method in use ('gssapi-with-mic')
+	Method string
+}
+
+func (mic *GSSAPIMicField) unmarshal(b []byte) error {
+	err := ssh.Unmarshal(b, mic)
```
Comment (janosdebugs) [inline internal/auth/client_kerberos.go:320]: We are introducing a dependency on the unstable SSH API here. I would strongly suggest we move this into a wrapper function in the SSH package or come up with a better API that does not require dependence on the SSH package.

File: internal/auth/client_kerberos.go [hunk 16]
```diff
@@ -0,0 +1,477 @@
...
+	var field GSSAPIMicField
+	err = field.unmarshal(micField)
+	if err != nil {
+		return err
+	}
+
+	if field.Request != SSH_MSG_USERAUTH_REQUEST || field.Service != "ssh-connection" || field.Method != "gssapi-with-mic" {
+		return message.NewMessage(
+			message.EAuthKerberosVerificationFailed,
+			"Verify() returned unverified but no error",
```
Comment (janosdebugs) [inline internal/auth/client_kerberos.go:356]: This error message does not seem to be very helpful. Maybe we should improve it to help with debugging?

File: internal/auth/client_kerberos.go [hunk 17]
```diff
@@ -0,0 +1,477 @@
...
+		)
+	}
+
+	if k.client.config.EnforceUsername && field.UserName != k.username {
+		return message.NewMessage(
+			message.EAuthKerberosVerificationFailed,
+			"Verify() returned unverified but no error",
```
Comment (janosdebugs) [inline internal/auth/client_kerberos.go:363]: Same as above, plus we should give the user a helpful error message that the username is the problem. Let's use `message.UserMessage` instead.

File: internal/auth/client_kerberos.go [hunk 18]
```diff
@@ -0,0 +1,477 @@
...
+func (k *kerberosAuthContext) tryAllowLogin(username  string) (bool ,error) {
+	authzRequest := auth.AuthorizationRequest{
+		PrincipalUsername: k.username,
+		LoginUsername: username,
+		RemoteAddress: k.remoteAddr.String(),
+		ConnectionID: k.connectionId,
+	}
+	authzResponse := &auth.ResponseBody{}
+
+	statusCode, err := k.client.http.Post("/authz", authzRequest, authzResponse)
```
Comment (janosdebugs) [inline internal/auth/client_kerberos.go:385]: We should make authz generally available for all auth methods.

File: internal/auth/client_kerberos_test.go [hunk 19]
```diff
@@ -0,0 +1,355 @@
...
+	"github.com/containerssh/gokrb5/v8/crypto"
+	"github.com/containerssh/gokrb5/v8/gssapi"
+	"github.com/containerssh/gokrb5/v8/iana/flags"
+	"github.com/containerssh/gokrb5/v8/messages"
+	"github.com/containerssh/gokrb5/v8/spnego"
+	"github.com/containerssh/gokrb5/v8/types"
+)
+
+func tempFile() *os.File {
+	file, err := ioutil.TempFile("", "krb5.keytab-*")
```
Comment (janosdebugs) [inline internal/auth/client_kerberos_test.go:32]: Please use `t.TempDir()` for the directory.

File: internal/auth/client_kerberos_test.go [hunk 20]
```diff
@@ -0,0 +1,355 @@
...
+		),
+		gssClient,
+	)
+	if err == nil {
+		assert.Fail(t, "Invalid username was successfully authenticated")
+	}
+}
+
+
+// Partially from kerberosAuthClient::Password
```
Comment (janosdebugs) [inline internal/auth/client_kerberos_test.go:183]: What is the source of this? Do we need to consider licensing issues here?

File: internal/auth/dist/CHANGELOG.md [hunk 21]
```diff
@@ -0,0 +1,5 @@
+## Changelog
```
Comment (janosdebugs) [inline internal/auth/dist/CHANGELOG.md:1]: This file seems to be unnecessary for the merge unless we want to introduce per-package changelogs.

File: internal/auth/testdata/krb5.conf [hunk 22]
```diff
@@ -0,0 +1,20 @@
+[logging]
+ default = FILE:/var/log/krb5libs.log
```
Comment (janosdebugs) [inline internal/auth/testdata/krb5.conf:2]: This may not work on all setups / Windows. I would suggest disabling logging for the test environment.

File: internal/docker/docker_impl.go [hunk 23]
```diff
@@ -358,6 +358,69 @@ loop:
 	return err
 }
 
+func (d *dockerV20Container) writeFile(path string, content []byte) error {
+	if d.config.Execution.Mode == config.DockerExecutionModeSession && d.config.Execution.DisableAgent {
+		return message.NewMessage(
+			message.EDockerWriteFileFailed,
+			"Agent needs to be enable in order to write files",
```
Comment (janosdebugs) [inline internal/docker/docker_impl.go:365]: `"The ContainerSSH guest agent is disabled and session mode is active. Failed to write file %s in the container."`

File: internal/docker/docker_impl.go [hunk 24]
```diff
@@ -358,6 +358,69 @@ loop:
 	return err
 }
 
+func (d *dockerV20Container) writeFile(path string, content []byte) error {
+	if d.config.Execution.Mode == config.DockerExecutionModeSession && d.config.Execution.DisableAgent {
```
Comment (janosdebugs) [inline internal/docker/docker_impl.go:362]: Does writing files work if the agent is disabled in connection mode?

File: internal/sshserver/dist/CHANGELOG.md [hunk 25]
```diff
@@ -0,0 +1,5 @@
+## Changelog
```
Comment (janosdebugs) [inline internal/sshserver/dist/CHANGELOG.md:1]: This changelog seems to be unnecessary.

File: internal/test/krb/root/usr/local/bin/init.sh [hunk 26]
```diff
@@ -25,6 +25,18 @@ kadmin.local -q "addprinc -pw ${KERBEROS_PASSWORD} ${KERBEROS_USERNAME}"
 echo -n "" >/etc/krb5kdc/kadm5.acl
 echo "${KERBEROS_USERNAME}@TESTING.CONTAINERSSH.IO *" >>/etc/krb5kdc/kadm5.acl
 
+echo -e "\e[32mAdding host principal testing.containerssh.io ...\e[0m"
+kadmin.local -q "addprinc -randkey host/testing.containerssh.io"
```
Comment (janosdebugs) [inline internal/test/krb/root/usr/local/bin/init.sh:29]: Ideally, we wouldn't hard-code credentials but pass them via environment variables like above. However, I can live with this.

File: message/auth.go [hunk 27]
```diff
@@ -96,3 +96,15 @@ const EAuthDeviceFlowRateLimitExceeded = "OAUTH_DEVICE_FLOW_RATE_LIMIT_EXCEEDED"
 // EAuthGitHubUsernameDoesNotMatch indicates that the user specified a username other than their GitHub login and
 // enforceUsername was set to on.
 const EAuthGitHubUsernameDoesNotMatch = "GITHUB_USERNAME_DOES_NOT_MATCH"
+
+// EAuthKerberosVerificationFailed indicates that there was an error verifying the kerberos ticket sent by the client
+const EAuthKerberosVerificationFailed = "KRB_VERIF_ERROR"
```
Comment (janosdebugs) [inline message/auth.go:101]: `KRB_VERIFY_ERROR`?

File: internal/auth/client_kerberos.go [hunk 28]
```diff
@@ -0,0 +1,477 @@
...
+			k.username,
+			username,
+		)
+	}
+	if !k.client.config.Authz.Enable {
+		// Allow all since authz server is disabled
+		return nil
+	}
+
+	retrier := retry.New(
```
Comment (janosdebugs) [inline internal/auth/client_kerberos.go:428]: Can we move adding retries to a separate PR and do it the manual way here? I'd rather introduce retries globally in a single PR, that would make it easier to add customizations such as logging.

**Review by janosdebugs — APPROVED** [review]
Awesome work, only a few minor things to change.

File: internal/auth/client_kerberos.go (hunk 17 above)
Comment (tsipinakis) [reply internal/auth/client_kerberos.go:361]: Will update the message as I forgot to updated from copy-pasting from above. However, is there a point in using `UserMessage`? There is no mechanism to return error messages in the ssh protocol?

File: internal/sshserver/dist/CHANGELOG.md (hunk 25 above)
Comment (tsipinakis) [reply internal/sshserver/dist/CHANGELOG.md]: This is autogenerated from goreleaser, was added accidentally. Whoops.

File: internal/docker/docker_impl.go (hunk 24 above)
Comment (tsipinakis) [reply internal/docker/docker_impl.go]: I'm not sure how that condition made it in there. Fixed.

File: internal/auth/client_kerberos_test.go (hunk 20 above)
Comment (tsipinakis) [reply internal/auth/client_kerberos_test.go]: I meant that it was initially taken from `client_kerberos.go` the password auth routine. Though after many iterations it's basically unrecognizable. Removed the comment.

File: internal/auth/client_kerberos.go (hunk 18 above)
Comment (tsipinakis) [reply internal/auth/client_kerberos.go]: Done!

File: auth/protocol.go [hunk 29]
```diff
@@ -60,6 +60,60 @@ type PublicKeyAuthRequest struct {
...
+	// backend as environment variables.
+	Metadata    map[string]string `json:"metadata,omitempty"`
+	// Files is a key-value pair of files to be placed inside containers.
+	// The key represents the path to the file while the value is the
+	// binary content.
+	Files       map[string][]byte `json:"files,omitempty"`
+	// Sensitivity is a set of key-value pairs that indicates the
+	// sensitivity of a value. The key identifies the entry name and is
+	// valid for all maps in ConnectionMetadata
+	Sensitivity map[string]int    `json:"sensitivity,omitempty"`
```
[the remark this answers is not in the document]
Comment (tsipinakis) [reply auth/protocol.go]: I ended up removing sensitivity entirely. Since the credentials are a file now only `Metadata` is send to the webhooks, environment and files stay only in the server and are send to the container at the end of the chain.

File: auth/webhook/client.go [hunk 30]
```diff
@@ -3,6 +3,7 @@ package webhook
 import (
 	"net"
 
+	auth2 "github.com/containerssh/libcontainerssh/auth"
```
[the remark this answers is not in the document]
Comment (tsipinakis) [reply auth/webhook/client.go:6]: This is following the existing naming used in other parts of the project (grep for `config2`). It should probably be changed globally later.

**Comment by tsipinakis** [comment]
@janosdebugs Ready for another round of reviews :)

File: config/auth.go [hunk 31]
```diff
@@ -293,3 +300,64 @@ func (o *AuthOIDCConfig) Validate() error {
...
+	// correctly this should match the contents of the KRB5CC environment
+	// variable.
+	CredentialCachePath string `json:"credentialCachePath" yaml:"credentialCachePath" default:"/tmp/krb5cc"`
+	// AllowPassword controls whether kerberos-based password
+	// authentication should be allowed. If set to false only GSSAPI
+	// authentication will be permitted
+	AllowPassword bool `json:"allowPassword" yaml:"allowPassword" default:"true"`
+	// ConfigPath is the path of the kerberos configuration file. This is
+	// only used for password authentication.
+	ConfigPath string `json:"configPath" yaml:"configPath" default:"/etc/containerssh/krb5.conf"`
```
[the remark this answers is not in the document]
Comment (tsipinakis) [reply config/auth.go:357]: KRB config is quite complex, I'd prefer not to at least until we have a compelling reason to make the change.
Additionally, this allows using the distribution/packaged config directly instead of translating it to containerssh cofig, which may be useful in some cases.

File: auth/protocol.go [hunk 32]
```diff
@@ -60,6 +60,60 @@ type PublicKeyAuthRequest struct {
...
+
+type ConnectionMetadata struct {
+	// Metadata is a set of key-value pairs that can be returned and
+	// either consumed by the configuration server or exposed in the
+	// backend as environment variables.
+	Metadata    map[string]string `json:"metadata,omitempty"`
+	// Files is a key-value pair of files to be placed inside containers.
+	// The key represents the path to the file while the value is the
+	// binary content.
+	Files       map[string][]byte `json:"files,omitempty"`
```
[the remark this answers is not in the document]
Comment (tsipinakis) [reply auth/protocol.go:104]: Added!

File: internal/auth/client_kerberos.go [hunk 33]
```diff
@@ -348,17 +345,19 @@ func (k *kerberosAuthContext) VerifyMIC(micField []byte, micToken []byte) error
 		return err
 	}
 
-	if field.Request != SSH_MSG_USERAUTH_REQUEST || field.Service != "ssh-connection" || field.Method != "gssapi-with-mic" {
+	if field.Request != internalSsh.SSH_MSG_USERAUTH_REQUEST || field.Service != "ssh-connection" || field.Method != "gssapi-with-mic" {
 		return message.NewMessage(
 			message.EAuthKerberosVerificationFailed,
-			"Verify() returned unverified but no error",
+			"Received MIC packet with unexpected values",
```
Comment (janosdebugs) [inline internal/auth/client_kerberos.go:351]: This error message is not very helpful, we should include the values we received if possible.

File: auth/protocol.go [hunk 34]
```diff
@@ -91,12 +91,12 @@ type AuthorizationRequest struct {
...
-	// METADATA_INTERAL signifies that a value is internal and for use only in containerssh
-	METADATA_INTERNAL     = 0
-	// METADATA_CONFIG signifies that a value can be exposed to the configuration server
-	METADATA_CONFIG       = 1 << 0
-	// METADATA_CONTAINER signifies that a value can be exposed to the pod as environment variables or files
-	METADATA_CONTAINER    = 1 << 1
+	// MetadataInternal signifies that a value is internal and for use only in containerssh
+	MetadataInternal     = 0
+	// MetadataConfig signifies that a value can be exposed to the configuration server
+	MetadataConfig       = 1 << 0
```
Comment (janosdebugs) [inline auth/protocol.go]: It would be better to use `iota` here.

**Review by janosdebugs — CHANGES_REQUESTED** [review]
@tsipinakis a few minor changes.

File: config/auth.go [hunk 35]
```diff
@@ -293,3 +309,58 @@ func (o *AuthOIDCConfig) Validate() error {
...
+	// authenticated user matches the SSH username entered. If set to false
+	// the authorization server must be responsible for ensuring proper
+	// access control.
+	//
+	// WARNING: If authorization is unset and this is set to false all
+	// authenticated users can log in to any account!
+	EnforceUsername bool `json:"enforceUsername" yaml:"enforceUsername" default:"true"`
+	// CredentialCachePath is the path in which the kerberos credentials
+	// will be written inside the user containers.
+	CredentialCachePath string `json:"credentialCachePath" yaml:"credentialCachePath" default:"/tmp/krb5cc"`
```
[the remark this answers is not in the document]
Comment (tsipinakis) [reply config/auth.go:350]: It does, the agent write-file command currently only writes with `0600` permissions. 
https://github.com/ContainerSSH/agent/pull/3/files#diff-0965f1fa88ac40fcfc62a7eb27c953cf287c8d9a10e80f5d6e84aeb165322594R15

File: internal/auditlogintegration/integration_test.go [hunk 36]
```diff
@@ -209,10 +211,10 @@ func (b *backendHandler) OnAuthKeyboardInteractive(
 		questions sshserver.KeyboardInteractiveQuestions,
 	) (answers sshserver.KeyboardInteractiveAnswers, err error),
 	_ string,
-) (response sshserver.AuthResponse, metadata map[string]string, reason error) {
+) (response sshserver.AuthResponse, metadata *auth2.ConnectionMetadata, reason error) {
 	answers, err := challenge(
 		"Test",
-		sshserver.KeyboardInteractiveQuestions{{
+		sshserver.KeyboardInteractiveQuestions{sshserver.KeyboardInteractiveQuestion{
```
[the remark this answers is not in the document]
Comment (tsipinakis) [reply internal/auditlogintegration/integration_test.go:217]: I'm pretty sure I made this one because golang-ci complained about it, I'll revert and test

File: auth/protocol.go (hunk 34 above)
Comment (tsipinakis) [reply auth/protocol.go]: This was removed in `077623b`, did you review an older revision?

File: internal/auth/client_kerberos.go [hunk 37]
```diff
@@ -34,7 +32,8 @@ type kerberosAuthContext struct {
 	remoteAddr net.IP
 
 	key types.EncryptionKey
-	username string
+	princUsername string
```
[the remark this answers is not in the document]
Comment (tsipinakis) [reply internal/auth/client_kerberos.go]: I'm using vim so I don't :smile: , but being verbose is probably better indeed.

File: auth/protocol.go (hunk 34 above)
Comment (janosdebugs) [reply auth/protocol.go]: Possibly, I went through the commits.

**Review by janosdebugs — APPROVED** [review]

**Review by janosdebugs — APPROVED** [review]
